### ISSUE TITLE
Is LCB actually a benefit in self-play?

### ISSUE THREAD

**Issue by lightvector** [leela-zero/leela-zero#2411]
Motivated a little by the mention of LCB in #2410 (which of course may or may not actually have anything to do with LCB, but nonetheless), I suppose I should go ahead and share some puzzling observations I've found in KataGo.

![image](https://user-images.githubusercontent.com/11942395/58662530-6c2e8880-82f8-11e9-8cc8-876c507a2633.png)

This is a graph of Elo improvement versus training steps for 4 partial runs of KataGo, where I forked a test run that was partway through training at 15-block net size into 4 parallel runs with different use of LCB.  Elo was evaluated separately from the training or gating by playing matches between various differently separated versions. Matches were performed without LCB, so as to only try to evaluate the raw neural net strength.
* LCBMoves: use LCB to select moves in self-play games
* LCBPolicy: use LCB to modify the numbers of visits written down for policy training. If a move is LCB winner but not max-visits move, then the number of recorded visits for the purpose of policy training is equal to the number of visits on the max visits move times the factor by which LCB winning move exceeds the minimum number of visits for the LCB formula to make it the winning move. So for example, if the the LCB winner got 150 visits and the max visits move was 200, but the LCB winner would have still been (barely) a winner given its current MCTS value at 100 visits, then factor is 1.5 and it gets record as "300 visits" for the purposes of policy training.
* LCBMoves+Policy: do both of the above

The graph is a little noisy, but notably, both forks with LCBMoves did slightly _worse_ than both forks without LCBMoves, and were both behind baseline for almost the entire duration. This is puzzling, as one would expect LCB to improve the playing strength, thereby getting more accurate game results for value head training. I definitely did not expect this when running this experiment.

Both forks with LCBPolicy finished slightly ahead of the corresponding run that was the same but without LCBPolicy. So maybe LCBPolicy is a good idea, but it's clear from the graph that it wasn't consistent through the duration of the runs, so I would say it's still unclear.

I was also experimenting with learning rate at the same time, so here is baseline along with two more forks run at double learning rate with and without LCBMoves:
![image](https://user-images.githubusercontent.com/11942395/58663922-874ec780-82fb-11e9-965a-bd8bb6660959.png)

LCBMoves also falls well behind here, although of course there is evidently some variability and noise across time.

So it seems like at least in KataGo, LCB seems not to be a clear benefit for training, and if anything, LCBMoves perhaps makes it worse. This is despite the fact that I've already verified with the same versions of KataGo that LCB easily adds +100 Elo or more at similar numbers of visits! (Settings are to use 4 stdevs with t vs normal correction, equivalent to a naive p value of about 3e-5, requiring also the LCB winner to at least have 5% of the visits of the max visits move).

One explanation I thought of was that maybe since LCB picks moves with lower visits sometimes, maybe it actually harms the strength "per amount of computation" because the next search will have less NN cache and effectively less tree reuse. And in selfplay training, strength "per computation" is very relevant. But I've also tested Elo strength at fixed playouts rather than visits and LCB clearly and greatly improves match strength there as well. 

So I don't really understand why, but these are the results I'm getting.

KataGo is pretty similar to LZ in the fundamental way that the self play works and how MCTS and policy target are used and are trained, although other parts are not the same, due to having additional board features, max visits variation, and additional training targets, etc. It would be very interesting therefore to know whether this counter-intuitive result holds up for LZ. Maybe attempts to replicate in MiniGo or some 9x9 experiments could help shed some light? Because it it does hold up, then possibly LCB is actually slowing LZ's learning, or at least not helping.

**Comment by alreadydone** [comment]
My thoughts:

+ It could be that LCBMove plays too conservatively, which may be good in matches but makes risky killer moves with higher averaged winrate under-explored in self-play games. LCBPolicy, IIUC, just marks the LCB winner as a promising move (more than the highest visited move) to search,  and doesn't quite hamper the diversity and extremeness of moves explored. LCBMove allows surprising moves (having lower prior) to be picked before visits reach the top, but only when the winrates in the subtree are consistently higher.

+ To be fair, shouldn't you do matches with LCB move selection strategy as well? After all, the trained winrate estimation is conditioned on that both sides make moves under the self-play setting (including noise, randomness, MCTS readouts, and also move selsction strategy).

**Comment by hzyhhzy** [comment]
Are there any experiments on different playouts?

**Comment by hzyhhzy** [comment]
I'm more curious about what will happen if we use 0.5x playouts or 2x playouts

**Comment by iopq** [comment]
@hzyhhzy since it's better to sample moves from games instead of using every move of the game, just use 1600 playouts for moves you will learn from, and like 400 for moves in between.

I believe KataGo already does this successfully

**Comment by hzyhhzy** [comment]
What will happen if use 200po for moves that will learn,and 50po for others? My experiments shows that 10po seems to be enough for 9x9, and at most 70po for 13x13

**Comment by l1t1** [comment]
i suggest increase visits to play more high quality games to learn

**Comment by iopq** [comment]
@hzyhhzy I don't know the exact values, but if you take only a few moves from each game, you can spend a lot of visits on the ones you learn on. Each game is around 200 moves, so even if 20 moves run at 5000 visits and the rest at 200 that's still fewer visits than currently

**Comment by hzyhhzy** [comment]
I mean:What if using 1/4 playouts?
What will happen if I use 100po for 19x19?Will be stuck or progress much faster?

**Comment by lightvector** [comment]
@alreadydone - Good point about wanting to also use match games with LCB to compare as well to be fair. I'm starting some matches between all these versions with LCB enabled now. I suspect it shouldn't affect the results much, but we'll see.

@hzyhhzy @l1t1  - I don't see what increasing or decreasing playouts or visits has to do with the topic of this issue. The potential point for this issue here is - is the well-intended change to enable LCB in self-play a few months back potentially harming Leela Zero? Because in KataGo, I have not found a way yet to do LCBMoves without harming the rate of self-play learning. Whether you should overall increase or decrease visits or playouts is certainly worth discussing, but seems like that should be in a separate issue.

**Comment by Vandertic** [comment]
@lightvector at what point of the run did you fork? I mean, what was roughly the level of play? Still bad/decent/amateur/pro/superhuman?

**Comment by lightvector** [comment]
Very roughly, the range of the graph goes from around LZ115 or LZ120 strength up to LZ150ish*. The neural net size is 15 blocks 192 channels. So, probably pro up through superhuman.

(* Yes, the test runs here peak out at a strength much greater than the peak strength reached by the 'official' released run of KataGo).

(Edit: A little further clarification, the fork itself took place near the left edge of the graph, just a little after 100M training steps, a little off the bottom of the graph here. In case it's relevant, all these runs, including baseline actually use a higher learning rate than was used pre-fork, since by this point I had already found that the learning rate in the original run was too low and that a higher one was much better for reaching past LZ130, but I don't think that affects the analysis too much. And around 125M or so is when the training window should have stopped including any of the old pre-fork games)

**Comment by wonderingabout** [comment]
i think lz was already stalling just before LCB got implemented (arround lz203), then soon after LCB we saw many promotions on a row on what was a stalling lz
but then after some good contributions, we went back to stalling status again as it should have been without LCB

@alreadydone maybe this is just a timing thing, and LCB is stronger, and also @lightvector maybe katago data is a fluke or unsignificant

**Comment by Splee99** [comment]
This may be off topic. I found that using upper confident bound actually helps in handicap games where winrate is very low. So one option is to use lcb AND ucb in a game according to the winrate.

**Comment by tapsika** [comment]
Without anything tangible, I'm not surprised to see something that increases match strength but not learning strength. As I wrote a few times in the past these two are only loosely related.

One example is exploration. It is narrowed down to a level that maximises match strength - but it is quite obvious that a slightly weaker selfplay where exploration is significantly wider can push learning significantly faster. Not necessarily OC but quite possibly - and these things were never tested.

The net learns where search results differ from raw policy (a refined ladder guess, a new move etc). This factor ('s frequency) is ignored by max-strength settings, so those cannot guarantee max learn speed.

Edit: On closer look the effect seems stronger for value than for policy, so the above mentioned lcb training <> non-lcb match play (worse predictions) is likely a factor as well.

**Comment by lightvector** [comment]
Here are the analogous graphs for a set of about 45000 match games played with 800 visits between a wide variety of pairings between the exact same versions of the neural nets in the earlier graphs, but this time with LCB enabled on all matches with parameters equal to the ones used in training. Elo ratings computed via Bayesian maximum likelihood over all the new games, as before:

![image](https://user-images.githubusercontent.com/11942395/58754182-d120e480-8498-11e9-9ad9-fcdeb4e3baf5.png)
![image](https://user-images.githubusercontent.com/11942395/58754183-db42e300-8498-11e9-98f7-02dda57dc3f5.png)

There are some differences, but they look pretty similar, with LCBMoves nets still comparing unfavorably overall.

@wonderingabout 
> i think lz was already stalling just before LCB got implemented (arround lz203), then soon after LCB we saw many promotions on a row on what was a stalling lz
> but then after some good contributions, we went back to stalling status again as it should have been without LCB

Given that LZ training seems noisy and has often had periods of slower and faster learning in the past (and many, many cases of people claiming stalls at various points before), the obvious question is - how should one tell the difference between the possibility you suggest, and the possibility that the seeming variation in progress rate is just noise and it happened in spite of LCB rather than because of it?

@alreadydone 
> * It could be that LCBMove plays too conservatively, which may be good in matches but makes risky killer moves with higher averaged winrate under-explored in self-play games. LCBPolicy, IIUC, just marks the LCB winner as a promising move (more than the highest visited move) to search,  and doesn't quite hamper the diversity and extremeness of moves explored. LCBMove allows surprising moves (having lower prior) to be picked before visits reach the top, but only when the winrates in the subtree are consistently higher.

@tapsika 
> Without anything tangible, I'm not surprised to see something that increases match strength but not learning strength. As I wrote a few times in the past these two are only loosely related.
> 
> One example is exploration. It is narrowed down to a level that maximises match strength - but it is quite obvious that a slightly weaker selfplay where exploration is significantly wider can push learning significantly faster. Not necessarily OC but quite possibly - and these things were never tested.

Thanks for the thoughts! Yeah, I agree, this isn't *extremely* surprising. But I still do find it surprising. Anecdotally when I look at a position where the LCB winner is different than the max visits, the vast majority of time LCB winner is a low-policy-prior move that did well compared to the max visits move and other high policy prior moves. So it seems to me like LCB should be resulting in *more* exploration not less, because games are more likely to take low-policy paths that the neural net didn't "expect" to happen.

Another intuition comes from this observation: LCB basically chooses moves by max(value - c * sigma / sqrt(visits)), were c might vary due to t-distribution vs normal, but in any case is a small constant. **When the relevant moves have fairly similar sigma, max_visits can be seen as the same as LCB where c = infinity!** Because choosing the move with the highest visits is choosing the move with the narrowest confidence interval, regardless of where its mean is. So LCB seems like it behaves *less conservatively* rather than more, because it merely needs 4 or 5 sigmas of confidence for value to override visits, rather than infinitely many (i.e. not allowing it to override at all).

Of course, that's for the case where "all relevant moves have fairly similar sigma", which probably often isn't true. When they are different, obviously LCB will preference the ones with smaller sigma as well, which should also lead to avoiding risks too. So perhaps this sigma variation effect outweighs the effect above that should increase exploration, while also being subtler and harder to notice anecdotally, leading to what @alreadydone suggested. Or maybe it's some other effect entirely.

I'll probably do more testing on this in the future, including maybe also a variation of LCB that is more like simply max( value - c / sqrt(visits) ) to eliminate the effect of sigma. It might take a while though to get to it though - even in KataGo, new or forked runs are costly and take some time to run.

Curious if anyone else has thoughts about what if anything can be done in LZ. To my knowledge, the way LCB is being used in LZ right now is pretty much the same as LCBMoves, which is concerning. LZ and KataGo are different enough to perhaps want to consider further testing before simply generalizing this result, which is why I was wondering about what other experiments could be done closer to the LZ framework to provide additional data points.

**Comment by wonderingabout** [comment]
@lightvector 
if one is truly interested in scientific research on this question, i think it is possible to resume training from lz203 but never implementing lcb ever (using same software version 0.16), and see if the curve shows the same behaviour than with lcb or flattens as i am thinking
one may repeat this experiment 3 or 5 times to gain some significance

**Comment by Splee99** [comment]
@wonderingabout I bet the curve is flat for sure. To my understanding LCB provided an opportunity for the training to follow a slightly different path to reach another optimum. I would think another parameter change in the UCT search in selfplay games may do the same thing.

**Comment by Splee99** [comment]
On the other hand, if the training is so closely related to the UCT search, the 15b net has certainly more advantage in time parity games than the 40b net because it makes more visits in the same time.

**Comment by tapsika** [comment]
@lightvector 
>So it seems to me like LCB should be resulting in more exploration not less, because games are more likely to take low-policy paths that the neural net didn't "expect" to happen.

I agree (as far as this is guessable), but policy in you graphs doesn't really seem inferior to baseline, is it? The effect seems to be about value mostly.

Actually (I'm NOT suggesting this seriously, the whole thing is too unclear to guess) *if* LCB really means more exploration, that offers another possible factor. Maybe widely explored games are simply more unpredictable than narrower ones. So value learns slower because the task is harder.

**Comment by tapsika** [comment]
Some random thoughts pushing these baseless theories a bit further:

IIUC the slower Elo gain was measured against fixed earlier versions? I wonder if the results would be the same against external anchors like Elf. Wider exploration (if this assumption is correct) might reduce the accuracy of predicting the behaviour and chances (thus outcome/value) of the same player, but at the same time might increase the overall accuracy for unknown players.

Even if the effect would remain (as probably a bit more likely) there may be more to this. It also seems possible that just predicting one side more accurately is very valuable even at the cost of more inaccuracy over the other side. So the player can select positions/lines better knowing at least what it can expect/do with them for itself. EDIT: Or another interpretation: it may be more valuable to predict what can be predicted more accurately (a narrow player's expectation) - that may allow better selection even if it's worse match for the actual players' behaviour.

The point here is that - if all these wild assumptions hold - it seems possible that training value on the same games as policy will always hold back growth, as ideally these would require significantly different training data (wider exploration for policy, narrow "rollouts" for value - and probably some different third settings for actual play, since lcb for example also gives higher Elo by itself).

(And iirc original AG found value net to be comparably useful as old fashioned half-dumb rollouts, this also points in a similar direction.)
